src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -179,12 +135,128 @@ int MPL_gpu_free(void *ptr)
     return MPL_ERR_GPU_INTERNAL;
 }
 
-int MPL_gpu_init()
+int MPL_gpu_init(int local_size, int local_rank, int need_thread_safety)
 {
-    return MPL_SUCCESS;
+    int i, mpl_err = MPL_SUCCESS;
+
+    node_local_size = local_size;
+    self_local_rank = local_rank;
+    cuda_ipc_handle_trees = MPL_malloc(sizeof(MPL_gavl_tree_t) * local_size, MPL_MEM_OTHER);
+    assert(cuda_ipc_handle_trees != NULL);
+    for (i = 0; i < local_size; ++i) {
+        mpl_err =
+            MPL_gavl_tree_create(sizeof(gpu_ipc_handle_obj_s), need_thread_safety,
+                                 &cuda_ipc_handle_trees[i]);
+        if (mpl_err != MPL_SUCCESS) {
+            MPL_gpu_finalize();
+            break;
+        }
+    }
+    return mpl_err;
 }
 
 int MPL_gpu_finalize()
 {
+    if (cuda_ipc_handle_trees) {
+        int i;
+        for (i = 0; i < node_local_size; ++i)
+            if (cuda_ipc_handle_trees[i])
+                MPL_gavl_tree_free(cuda_ipc_handle_trees[i], gpu_ipc_handle_free);
+    }
+    MPL_free(cuda_ipc_handle_trees);
+    return MPL_SUCCESS;
+}
+
+static void gpu_ipc_handle_free(void *handle_obj)
+{
+    gpu_ipc_handle_obj_s *handle_obj_ptr = (gpu_ipc_handle_obj_s *) handle_obj;
+    cudaIpcCloseMemHandle((void *) ((char *) handle_obj_ptr - handle_obj_ptr->offset));
+    MPL_free(handle_obj);
+    return;
+}
+
+int MPL_gpu_ipc_handle_create(const void *addr, MPL_gpu_ipc_mem_handle_t * ipc_handle)
+{
+    cudaError_t ret;
+    CUresult curet;
+    CUdeviceptr pbase;
+    size_t len;
+    struct cudaPointerAttributes ptr_attr;
+
+    curet = cuMemGetAddressRange(&pbase, &len, (CUdeviceptr) addr);
+    CU_ERR_CHECK(curet);
+
+    ipc_handle->remote_base_addr = (uintptr_t) pbase;
+    ipc_handle->len = len;
+    ipc_handle->node_rank = self_local_rank;
+    ret = cudaPointerGetAttributes(&ptr_attr, addr);
+    CUDA_ERR_CHECK(ret);
+    ipc_handle->dev_id = (int) ptr_attr.device;
+
+    ret = cudaIpcGetMemHandle(&ipc_handle->handle, (void *) addr);
+    CUDA_ERR_CHECK(ret);
+
+    ipc_handle->offset = (uintptr_t) addr - (uintptr_t) pbase;
+
+  fn_exit:
     return MPL_SUCCESS;
+  fn_fail:
+    return MPL_ERR_GPU_INTERNAL;
+}
+
+int MPL_gpu_ipc_handle_map(MPL_gpu_ipc_mem_handle_t ipc_handle, void **addr)
+{
+    int mpl_err = MPL_SUCCESS;
+    int node_rank, found;
+
+    gpu_ipc_handle_obj_s *handle_obj =
+        (gpu_ipc_handle_obj_s *) MPL_malloc(sizeof(gpu_ipc_handle_obj_s), MPL_MEM_OTHER);
+    assert(handle_obj != NULL);
```

Comment:
Why are you not simply using a local stack variable for `handle_obj` instead of allocating it?

Reply:
`handle_obj` might be inserted into AVL, so it should be alive until tree is freed. It has to be malloc instead of local stack variable.

Reply:
Doesn't the MPL function copy the value in?

Reply:
no, it will just store a pointer to object.

Reply:
You copy-in just the pointer in the insert, but copy-out the full object in search?  That's not a good design.  It's inconsistent.

Reply:
so I want to let `search` to just return a pointer and delete `found` flag parameter. I feel this design is cleaner.

Reply:
Then the caller has to maintain all the objects and only free them at finalize.